lib/scanner/xml-scanner/xml-scanner.c

```diff
@@ -0,0 +1,383 @@
+  (void)self->text_function(self, element_name, text, text_len, error);
+}
+
+void
+xml_scanner_parse(XMLScanner *self, const gchar *input, gsize input_len, GError **error)
+{
+  g_assert(self->push_key_value.push_function);
+
+  GMarkupParser scanner_callbacks = { .start_element = _xml_scanner_start_element,
+                                      .end_element = _xml_scanner_end_element,
+                                      .text =  _xml_scanner_text
+                                    };
+  ScratchBuffersMarker marker;
+  scratch_buffers_mark(&marker);
+  self->xml_ctx = g_markup_parse_context_new(&scanner_callbacks, 0, self, NULL);
+  g_markup_parse_context_parse(self->xml_ctx, input, input_len, error);
+  if (error && *error)
+    goto exit;
+  g_markup_parse_context_end_parse(self->xml_ctx, error);
+
+exit:
+  scratch_buffers_reclaim_marked(marker);
+  g_markup_parse_context_free(self->xml_ctx);
+  self->xml_ctx = NULL;
+}
+
+
+/*
+  For some patterns, GPatternSpec stores the pattern as reversed
+  string. In such cases, the matching must be executed against the
+  reversed matchstring. If one uses g_pattern_match_string, glib will
+  reverse the string internally in these cases, but it is suboptimal if
+  the same matchstring must be matched against different patterns,
+  because memory is allocated each time and string is copied as
+  reversed, though it would be enough to execute this reverse once. For
+  that reason one can use g_pattern_match(), which accepts both the
+  matchstring and a reversed matchstring as parameters.
+
+  Though there are cases when no reverse is needed at all. This is
+  decided in the constructor of GPatternSpec, but at this time of
+  writing this information is not exported, and cannot be extracted in
+  any safe way, because GPatternSpec is forward declared.
+
+  This function below is an oversimplified/safe version of the logic
+  used glib to decide if the pattern will be reversed or not. One need
+  to note the worst case scenario is if syslog-ng will not reverse the
+  matchstring, though it should because in that case number of extra
+  memory allocations and string reversals scale linearly with the number
+  of patterns. We need to avoid not pre-reversing, when glib would.
+ */
+gboolean
+joker_or_wildcard(GList *patterns)
+{
+  gboolean retval = FALSE;
+  GList *pattern = patterns;
+  while (pattern != NULL)
+    {
+      gchar *str = ((gchar *)pattern->data);
+      if (strpbrk(str, "*?"))
+        {
+          retval = TRUE;
+          break;
+        }
+      pattern = g_list_next(pattern);
+    }
+
+  return retval;
+}
+
+static void
+_compile_and_add(gpointer tag_glob, gpointer exclude_patterns)
+{
+  g_ptr_array_add(exclude_patterns, g_pattern_spec_new(tag_glob));
+}
+
+void
+xml_scanner_init(XMLScanner *self, XMLScannerOptions *options, PushCurrentKeyValueCB push_function,
+                 gpointer user_data, gchar *key_prefix)
```

Comment:
For me `key_prefix` is an option-ish parameter. Is it possible to pass through XMLScannerOptions? `xml_scanner_set_key_prefix`?